2024/articles/omp-santa.pod

```diff
@@ -0,0 +1,339 @@
+
+        // Convert satellite's geographic coordinates to Cartesian coordinates
+        geo_to_cartesian(sat_latitudes[i], sat_longitudes[i], sat_altitudes[i], &sat_x, &sat_y, &sat_z);
+
+        // Calculate the distance from the sleigh to this satellite
+        _distances[i] = calculate_distance(sleigh_x, sleigh_y, sleigh_z, sat_x, sat_y, sat_z);
+    }
+  }
+
+=end c
+
+Jingles quickly typed up a solution in Perl, his favorite programming
+language. The solution was simple but critically offloaded the triangulation
+computations to C<Inline::C> code containing OpenMP directives, using the
+C<OpenMP> module on I<CPAN>.
+
+=begin perl
+
+  use strict;
+  use warnings;
+  
+  use OpenMP;
+     
+  use Inline (
+      C    => 'DATA',
+      with => qw/OpenMP::Simple/,
+  );
+     
+  my $omp = OpenMP->new;
+  
+  $omp->env->omp_num_threads($ARGV[0] // 8); # accept number of threads via commandline
+  
+  # Earth radius in meters
+  my $R = 6371000;
+  
+  # Sleigh's position (latitude, longitude, altitude)
+  my $sleigh_lat = 29.9500;
+  my $sleigh_lon = -90.070;
+  my $sleigh_alt = 300.0; # in meters
+  
+  # Satellite positions in tab-delimited format
+  my @satellites = ();
+  open my $FH, "<", "./satellite-data.txt" || die $!;
+  
+  foreach my $line (<$FH>) {
+    chomp $line;
+    push @satellites, [split(/[\s\t]+/, $line)];
+  }
+  
+  # Function to calculate distance from sleigh to each satellite using OpenMP::Simple,
+  # a subclass of Inline::C!
+  my $distances = calculate_distances($sleigh_lat, $sleigh_lon, $sleigh_alt, \@satellites);
+  
+  # Print the calculated distances
+  foreach my $distance (@$distances) {
+      print "Distance: $distance meters\n";
+  }
+
+=end perl
+  
+Santa watched as Jingles split up the computational load. Each satellite’s
+data was handled in parallel by different threads, and within seconds, the
+recalculations were done. The GPS latency issues were fixed. Jingles already
+had some thoughts of improvements he could make using OpenMP's ability to
+target GPUs directly, but was quite happy with the current resolution.
+
+Just as the final calculations finished, a gentle voice spoke, “When Faith
+sanctifies morally neutral technology, Good things are possible at Internet
+scale.”
+
+Santa turned to see the Holy Family -  Baby Jesus in the arms of His Virgin
+Mother accompanied by His foster father, Joseph; by the sleigh, smiling
+serenely. “Thank you,” Santa whispered. “Merry Christmas.”
+
+With the sleigh back on track, Santa soared into the night sky.
+
+When things settled down, Santa was able to look at Jingles' full code, and
+it was something to behold!
+
+=begin perl
+
+  use strict;
+  use warnings;
+  
+  use OpenMP;
+     
+  use Inline (
+      C    => 'DATA',
+      with => qw/OpenMP::Simple/,
+  );
+     
+  my $omp = OpenMP->new;
+  
+  $omp->env->omp_num_threads($ARGV[0] // 8); # accept number of threads via commandline
+  
+  # Earth radius in meters
+  my $R = 6371000;
```

Comment:
This is not actually used, either here or above?

Reply:
```suggestion
```